factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/chains/TransformationChains.java

```diff
@@ -0,0 +1,161 @@
+import es.usc.citius.hipster.algorithm.Algorithm.SearchResult;
+import es.usc.citius.hipster.algorithm.Hipster;
+import es.usc.citius.hipster.graph.GraphBuilder;
+import es.usc.citius.hipster.graph.GraphSearchProblem;
+import es.usc.citius.hipster.graph.HipsterDirectedGraph;
+import es.usc.citius.hipster.model.impl.WeightedNode;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import lombok.Value;
+
+public class TransformationChains implements TransformationStoreListener {
+
+    private static final double BASE_COST = 1_000_000d;
+
+    private final SchemaRegistry r;
+
+    private final RegistryMetrics registryMetrics;
+
+    private final Map<TransformationKey, Map<VersionPath, TransformationChain>> cache = new HashMap<>();
+
+    @Value
+    static class VersionPath {
+        int fromVersion;
+
+        int toVersion;
+    }
+
+    public TransformationChains(SchemaRegistry r, RegistryMetrics registryMetrics) {
+        this.r = r;
+        this.registryMetrics = registryMetrics;
+        r.register(this);
+    }
+
+    @Value(staticConstructor = "of")
+    private static class Edge {
+        int fromVersion;
+
+        int toVersion;
+
+        Transformation transformation;
+
+        public static Edge from(Transformation t) {
+            return of(t.fromVersion(), t.toVersion(), t);
+        }
+    }
+
+    public TransformationChain get(TransformationKey key, int from, int to)
+            throws MissingTransformationInformation {
+
+        Map<VersionPath, TransformationChain> chainsPerKey;
+
+        synchronized (cache) {
+            // sync is necessary, because we don't want to end up with two
+            // different maps we're locking the whole cache, but creating a
+            // hashmap should not take too much time
+            chainsPerKey = cache.computeIfAbsent(key, k -> new HashMap<>());
+        }
+        synchronized (chainsPerKey) {
+            // we're locking the map for this particular key. contention should
+            // be limited and the gain of not creating unnecessary chains should
+            // be on the plus side.
+            return chainsPerKey.computeIfAbsent(new VersionPath(from, to), p -> build(key, from,
+                    to));
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private TransformationChain build(TransformationKey key, int from, int to)
+            throws MissingTransformationInformation {
+
+        GraphBuilder<Integer, Edge> builder = GraphBuilder.create();
+        List<Transformation> all = r.get(key);
+        if (all.isEmpty()) {
+            registryMetrics.count(MetricEvent.MISSING_TRANSFORMATION_INFO, Tags.of(
+                    Tag.of(RegistryMetrics.TAG_IDENTITY_KEY, key.toString()), Tag.of("from", String
+                            .valueOf(from)), Tag.of("to", String.valueOf(to))));
+
+            throw new MissingTransformationInformation("No Transformations for " + key);
```

Comment:
Throw an exception here means, that missing transformations are not treated as compatible (if I got it right ;-)). So doc has to be adjusted.

Reply:
this is intended. empty transformations are added when loading the registry data where applicable